xmountains.man
.de Id
.ds Rv \\$3
.ds Dt \\$4
..
.Id $Id: xmountains.man,v 1.14 1999/03/30 08:07:08 spb Exp $
.ds r \s-1RCS\s0
.if n .ds - \%--
.if t .ds - \(em
.TH XMOUNTAINS 1 
.SH NAME
xmountains \- A fractal landscape generator.
.SH SYNOPSIS
.B xmountains
[
.B bqgPdEmMrBnZIASTWFGCapcevfRltxsXYH
]

.SH DESCRIPTION
.B xmountains
is a X11 based fractal landscape generator.
It generates random fractal surfaces and displays them in a window.
While the program is running the fractal is continuously extended on the right
and the image is scrolled from right to left to expose the newly generated area.

.PP
Flags:
.br
.B \-b 
.RS
Use the root window.
.RE
.br
.B \-q 
.RS
Reset the root window when the program exits. This option is ignored
unless the program is using the root window.
.RE
.br
.B \-g [string] 
.RS
Set the geometry of the window.
.RE
.br
.B \-P [filename] 
.RS
Write the PID to the specified file.
.RE
.br
.B \-d [string] 
.RS
Set the display.
.RE
.br
.B \-E
.RS
Toggle the way that window repainting is done. By default the image
pixmap is installed as the background pixmap of the window and window
repainting should be performed automatically by the X server. If this
does not work on a particular system then this flag can be used to
enable explicit expose event processing. In the default mode root window images
will remain in place after the program exits.
.RE
.br
.B \-m 
.RS
Display a map of the surface rather than a perspective view.
.RE
.br
.B \-M
.RS
Disable reflections in the water.
.RE
.br
.B \-r [20] 
.RS
To reduce the load on the X server the program generates several columns of pixels before scrolling the image. This flag sets the number of columns.
A negative value reverses the direction of scrolling. A value of zero
makes the program scan across the window without scrolling at all.
.RE
.br
.B \-B [80] 
.RS
Set the number of shades of each colour that the program allocates.
.RE
.br
.B \-n [245] 
.RS
Set a maximum number of colours to use. This is just a different way of
setting the number of shades of each colour.
.RE
.br
.B \-Z [10] 
.RS
Set the number of seconds that the program sleeps after scrolling the screen.
.RE
.br
.B \-I [40.0] 
.RS
Set the vertical angle of illumination.
.RE
.br
.B \-A [0.0] 
.RS
Set the horizontal angle of illumination.
.RE
.br
.B \-S [0.6] 
.RS
Set the vertical stretch of the surface.
.RE
.br
.B \-T [0.5] 
.RS
Set the vertical shift of the surface.
.RE
.br
.B \-W [0.5] 
.RS
Set the height where the water occurs.
.RE
.br
.B \-F [1] 
.RS
Reduce the variation in the foreground height to ensure a good view of
the surface. This flag sets the number of iterations for which the
foreground height is constrained to a constant value.
.RE
.br
.B \-G [-1.0] 
.RS
Mean altitude for the for foreground.
.RE
.br
.B \-C [0.3] 
.RS
Set the contour parameter. The base colour of a point depends on its
height and how flat the surface is. This parameter controls the relative
importance of these factors.
.RE
.br
.B \-a [2.5] 
.RS
Set the Altitude of the viewpoint.
.RE
.br
.B \-p [4.0] 
.RS
Set the distance of the viewpoint from the front of the surface.
.RE
.br
.B \-c [1.0] 
.RS
Set the contrast parameter. Values too far from 1.0 will give strange results
.RE
.br
.B \-e [0.3] 
.RS
Set the ambient light level. This is a fractional value that sets how
bright shaded areas will be compared to fully illuminated ground.
.RE
.br
.B \-v [0.6] 
.RS
Set the vertical light level. The program also implements a secondary
light source shining from directly above the surface. This flag sets how
bright this light source is relative to the main light source.
.RE
.br
.B \-f [0.65] 
.RS
Set the fractal dimension of the surface. This should be in the range 0.5\-1.0.
.RE
.br
.B \-R [0] 
.RS
Set the seed for the random number generator. A value of 0 causes the seed to be set from the clock.
.RE
.br
.B \-l [10] 
.RS
Set the width of the surface. Increasing this value by one doubles the width of the surface.
.RE
.br
.B \-t [2] 
.RS
Set the number of non\-fractal iterations. Increasing this value by one doubles the average number of mountains across the width of the surface.
.RE
.br
.B \-x
.RS
Use cross updates (see under \-H)
.RE
.br
.B \-s [1]
.RS
Control the regeneration steps used to reduce creasing. There are three
possible regeneration steps controlled by the first 3 bits of this flag value.
(see under \-H)
.RE
.br
.B \-X [0.0]
.RS
regeneration step parameter (see under \-H)
.RE
.br
.B \-Y [0.0]
.RS
regeneration step parameter (see under \-H)
.RE
.br
.B \-window-id 0xNNNNN
.RS
draw on existing external window.
.RE
.br
.B \-K [1024]
.RS
initial width of window
.RE
.br
.B \-L [768]
.RS
initial height of window
.RE

.br
.B \-H
.RS
Print a short description of the algorithm explaining the \-x \-s \-X
and \-Y parameters.
.RE

.PP

.SH IDENTIFICATION
Author: S.Booth@epcc.ed.ac.uk





NEWS
CHANGES from V2.1
Added a -n flag to set the number of colours. This changes the same
parameter as -B but you request total number of colours.

CHANGES from V1.4
The fractal generation code has been reworked. There algorithm has been
extended significantly and should be able to generate much better
looking surfaces. There may be a performance hit for some options though.

CHANGES from V1.3
1) There is now a flag to control the horizontal light angle.
2) negative values of the scroll parameter make the image scroll in the
   opposite direction.
3) water reflections have been implemented.

CHANGES from V1.2
1) program now sleeps after a scroll rather than once per column.
2) added a -B flag to control the number of colours needed.
3) added a -q flag that clears the root window when the program exits.
4) fixed memory leak when in `map' mode.

CHANGES from V1.1
Several changes including:
1) There is now a secondary (vertical) light source to add some detail to 
the shadows.
2) There is a -Z flag to reduce the CPU load, once the initial screen has
been generated the program will call sleep once for each column of
pixels.
3) Xmountains creates a pixmap containing the current state of the image.
This is now installed as the background pixmap of the window. This means
that root-window images will remain in place if the program is stopped.
The -E flag disables this feature and explicitly traps expose events
instead. There may be some delay in repainting the window if both the -E
and -Z flags are used at the same time.


xmountains.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
#include "crinkle.h"
#include "paint.h"
#include "patchlevel.h"
#include "copyright.h"
#include<X11/Xlib.h>
#include<X11/Xutil.h>
#include<X11/Xatom.h>

#define VERSION 2
#define SIDE 1.0

char scroll_Id[]="$Id: xmountains.c,v 1.43 2010/12/17 09:25:36 spb Exp $";
extern Graph g;
Parm fold_param;
extern char *display;
extern char *geom;

extern int swosh;

/* {{{ my version on getopt*/
int optind=1;
char *optarg;
int opterr=1;

int my_getopt (int argc, char **argv, char *pat)
{
  char *flag;
  
  if((optind >= argc) || (argv[optind][0] != '-'))
  {
    return -1;
  }
  if( argv[optind][1] == '-' )
  {
    optind++;
    return -1;
  }
  if( argv[optind][1] == ':' )
  {
    if( opterr )
    {
      fprintf(stderr,"getopt: found \":\" in optstring\n");
    }
    return '?';
  }
  for(flag=pat;*flag;flag++)
  {
    if( *flag == argv[optind][1] )
    {
      optind++;
      if( *(flag+1) == ':' )
      {
        if(optind >= argc )
        {
          if( opterr )
          {
            fprintf(stderr,"getopt: no option for flag %c\n",*flag);
          }
          return '?';
        }
        optarg = argv[optind];
        optind++;
      }
      return *flag;
    }
      
  }
  if( opterr )
  {
    fprintf(stderr,"getopt: flag %s not recognized\n",argv[optind]);
  }
  optind++;
  return '?';
}
/* }}} */

void init_graphics (int, Window, int,int, Graph *, Gun *, Gun *, Gun *);
void clear_col( int );
void finish_graphics();
void plot_pixel (int, int, unsigned char);
void scroll_screen ( int );
void zap_events();

void finish_prog(int err);

int mapwid;

/* Go through argv and find and extract any "-window-id NNNN" or
   "-window-id 0xXXXX" option.  (Maybe this can be done with
   getopt, but I can't be bothered to figure it out.)
 */
static unsigned long window_id_kludge (int *argcP, char **argv)
{
  unsigned long id = 0;
  int i;
  int j;
  for (i = 1; i < *argcP; i++)
    {
      if (!strcmp (argv[i], "-window-id") && i+1 < *argcP)
        {
          char c;
          char *str = argv[i+1];

          if (1 != ((str[0] == '0' && (str[1] == 'x' || str[1] == 'X'))
                    ? sscanf (str+2, "%lx %c", &id, &c)
                    : sscanf (str,   "%lu %c", &id, &c)))
            continue;

          /* remove "-window-id xxx" from the arglist */
          for (j = i+2; j < *argcP; j++)
            argv[j-2] = argv[j];
          *argcP -= 2;
        }
    }
  return id;
}


void init_parameters();
void print_algorithm();
void seed_uni(int ijkl);
void plot_column(Graph *g);

/* Note, Robbie Hatley, 2024-07-16: When installing xmountains as an AUR package,
   "main" no-longer compiles without being declared "int", presumably due to
   recent changes in gcc. */
int main (int argc, char **argv)
{
  int i;
  int e_events=FALSE;
  int request_clear=FALSE;
  int smooth=1;
  int snooze=10;
  int root= 0;
  unsigned long window_id = 0;
  int seed=0;

  int c, errflg=0;
  extern char *optarg;
  extern int optind;
  char *mesg[2];
  Gun *clut[3];
  FILE *pidfile;
  struct sigaction sa;

  swosh = 0;          /* default is ! -w */

  init_parameters();
  /* {{{ handle command line flags*/

  window_id = window_id_kludge (&argc, argv);

  mesg[0]="false";
  mesg[1]="true";
  while((c = my_getopt(argc,argv,"whbxmqMEHl:r:f:t:I:A:S:T:W:C:a:p:B:n:R:g:d:c:e:v:Z:s:X:Y:P:F:G:K:L:"))!= -1)
  {
    switch(c){
       case 'w':
         swosh = TRUE;               /* update window only when pixmap is complete */
         break;
      case 'b':
        root = 1- root;
        break;                      /* run on root window */
      case 'x':
        fold_param.cross = 1- fold_param.cross;
        break;                      /* use cross updates */
      case 'E':
        e_events = 1 - e_events;
        break;
      case 'q':
        request_clear = 1 - request_clear;
        break;
      case 'm':                     /* Map view only */
        g.map = 1 - g.map;
        break;
      case 'M':                     /* put in reflections */
        g.reflec = 1 - g.reflec;
        break;
      case 'l':                     /* Set # levels of recursion */
         g.levels = atoi( optarg );
         if( g.levels < 2 )
         {
           g.levels = 2;
         }
         break;
      case 'F':                     /* Set # levels to force front to mean */
         fold_param.force_front = atoi( optarg );
         break;
      case 's':                     /* Set smoothing parameter */
         smooth = atoi( optarg );
         fold_param.rg1 = smooth & 1;
         fold_param.rg2 = smooth & 2;
         fold_param.rg3 = smooth & 4;
         break;
      case 't':                     /* Set width of lowest level */
         g.stop = atoi( optarg );
         if( g.stop < 0 )
         {
           g.stop = 0;
         }
         break;
      case 'r':
         g.repeat = atoi( optarg );
         if( g.repeat < 0 )
         {
           g.repeat = -g.repeat;
           i= -1;
         }else{
           i=1;
         }
         /* we want repeat to be a multiple of 2 as we are using
          * a textured field for the sky.
          */
         g.repeat = i*(2 * ((g.repeat +1)/2));
         break;
      case 'B':                     /* set band_size */
         g.band_size = atoi( optarg );
         if( g.band_size < 2 )
         {
           g.band_size=2;
         }
         g.n_col = (BAND_BASE + (N_BANDS * g.band_size));
         break;
      case 'n':                     /* set max number of colours */
         g.n_col = atoi( optarg );
         if( g.n_col < MIN_COL )
         {
           g.n_col = MIN_COL;
         }
         g.band_size = (g.n_col - BAND_BASE)/N_BANDS;
         g.n_col = (BAND_BASE + (N_BANDS * g.band_size));
         break;
      case 'R':                     /* set seed, read clock if 0 */
         seed = atoi( optarg );
         break;
      case 'Z':                     /* put sleep into wait events */
         snooze = atoi( optarg );
         if( snooze < 0 )
         {
           snooze = 0;
         }
         break;
      case 'P':
         pidfile = fopen(optarg,"w");
         if( pidfile )
         {
           fprintf(pidfile,"%d\n",getpid());
           fclose(pidfile);
         }else{
           perror(optarg);
         }
         break;
      case 'f':                     /* set fractal dimension */
         fold_param.fdim = atof( optarg );
         if( fold_param.fdim < 0.5 )
         {
          fold_param.fdim=0.5;
         }
         if( fold_param.fdim > 1.0 )
         {
          fold_param.fdim=1.0;
         }
         break;
      case 'I':                     /* set Illumination angle */
         g.phi = ((PI * atof( optarg ))/180.0);
         if ( g.phi < 0.0 )
         {
           g.phi=0.0;
         }
         if( g.phi > PI/2.0 )
         {
           g.phi = PI/2.0;
         }
         break;
      case 'A':                     /* set Illumination angle (horizontal)*/
         g.alpha = ((PI * atof( optarg ))/180.0);
         if( g.alpha < -PI/3.0 )
         {
           g.alpha = -PI/3.0;
         }
         if( g.alpha > PI/3.0 )
         {
           g.alpha = PI/3.0;
         }
         break;
      case 'X':                     /* set mix */
         fold_param.mix = atof( optarg );
         break;
      case 'Y':                     /* set midmix */
         fold_param.midmix = atof( optarg );
         break;
      case 'S':                     /* set stretch */
         g.stretch = atof( optarg );
         break;
      case 'W':                     /* set sealevel */
         g.sealevel = atof( optarg );
         break;
      case 'G':                     /* set forceheight */
         fold_param.forceval = atof( optarg );
         break;
      case 'T':                     /* set shift */
         g.base_shift = atof( optarg );
         break;
      case 'C':
         g.contour = atof( optarg );
         break;
      case 'a':                     /* set altitude */
         g.altitude = atof( optarg );
         break;
      case 'p':                     /* set distance */
         g.distance = atof( optarg );
         break;
      case 'c':
         g.contrast = atof( optarg );
         if( g.contrast < 0.0 )
         {
          g.contrast=0.0;
         }
         break;
      case 'e':
         g.ambient = atof( optarg );
         if( g.ambient < 0.0 )
         {
          g.ambient = 0.0;
         }
         if( g.ambient > 1.0 )
         {
          g.ambient=1.0;
         }
         break;
      case 'v':
         g.vfract = atof( optarg );
         if( g.vfract < 0.0 )
         {
          g.vfract = 0.0;
         }
         break;
      case 'g':
         geom = optarg;
         break;
      case 'd':
         display = optarg;
         break;
      case 'K':
	 g.graph_width = atoi( optarg );
	 if( g.graph_width < 10 )
	 {
	   g.graph_width=1024;
	 }
	 break;
      case 'L':
	 g.graph_height = atoi( optarg );
         if( g.graph_height < 10 ){
	   g.graph_height = 768;
	 }
	 break;
      case 'H':
         print_algorithm();
         errflg++;
         break;
      case 'h':
      case '?':
         errflg++;
    }
  }
  if( errflg )
  {
    fprintf(stderr,"%s: version %d.%d\n",argv[0],VERSION,PATCHLEVEL);
    fprintf(stderr,"usage: %s -[hwbqgdPEmMrBZIASFTCapcevfRltxsXYH]\n",argv[0]);
    fprintf(stderr," -h       Print this message\n");
    fprintf(stderr," -w       [%s] update window with complete images only\n",mesg[swosh]);
    fprintf(stderr," -b       [%s] use root window \n",mesg[root]);
    fprintf(stderr," -q       [%s] reset root window on exit\n",mesg[request_clear]);
    fprintf(stderr," -g string     window geometry\n");
    fprintf(stderr," -d string     display\n");
    fprintf(stderr," -P filename   write PID to file\n");
    fprintf(stderr," -E       [%s] toggle explicit expose events \n",mesg[e_events]);
    fprintf(stderr," -m       [%s] print map \n",mesg[g.map]);
    fprintf(stderr," -M       [%s] implement reflections \n",mesg[g.reflec]);
    fprintf(stderr," -r int   [%d] # columns before scrolling \n",g.repeat);
    fprintf(stderr," -B int   [%d] # shades in a colour band\n",g.band_size);
    fprintf(stderr," -n int   [%d] # number of colours\n",g.n_col);
    fprintf(stderr," -Z int   [%d] time to sleep before scrolling\n",snooze);
    fprintf(stderr," -I float [%f] vertical angle of light \n",(g.phi*180.0)/PI);
    fprintf(stderr," -A float [%f] horizontal angle of light \n",(g.alpha*180.0)/PI);
    fprintf(stderr," -S float [%f] vertical stretch \n",g.stretch);
    fprintf(stderr," -T float [%f] vertical shift \n",g.base_shift);
    fprintf(stderr," -W float [%f] sealevel \n",g.sealevel);
    fprintf(stderr," -F int   [%d] reduce variation in the foreground \n",fold_param.force_front);
    fprintf(stderr," -G float [%f] average foreground height \n",fold_param.forceval);
    fprintf(stderr," -C float [%f] contour parameter \n",g.contour);
    fprintf(stderr," -a float [%f] altitude of viewpoint \n",g.altitude);
    fprintf(stderr," -p float [%f] distance of viewpoint \n",g.distance);
    fprintf(stderr," -c float [%f] contrast\n",g.contrast);
    fprintf(stderr," -e float [%f] ambient light level\n",g.ambient);
    fprintf(stderr," -v float [%f] vertical light level\n",g.vfract);
    fprintf(stderr,"Fractal options:\n");
    fprintf(stderr," -f float [%f] fractal dimension \n",fold_param.fdim);
    fprintf(stderr," -R int   [%d] rng seed, read clock if 0 \n",seed);
    fprintf(stderr," -l int   [%d] # levels of recursion \n",g.levels);
    fprintf(stderr," -t int   [%d] # non fractal iterations \n",g.stop);
    fprintf(stderr," -x       [%s] cross update \n",mesg[fold_param.cross]);
    fprintf(stderr," -s       [%x] smoothing (0-7)\n",smooth);
    fprintf(stderr," -X float [%f] fraction of old value for rg2 & rg3\n",fold_param.mix);
    fprintf(stderr," -Y float [%f] fraction of old value for rg1\n",fold_param.midmix);
    fprintf(stderr," -window-id 0xNNNNN  draw on existing external window.\n");
    fprintf(stderr," -K int [%d] width of window\n",g.graph_width);
    fprintf(stderr," -L int [%d] height of window\n",g.graph_height);
    fprintf(stderr," -H            print short description of algorithm.\n");
    exit(1);
  }

  /* }}} */
  for(i=0 ;i<3 ;i++)
  {
    clut[i] = (Gun *) malloc(g.n_col * sizeof(Gun));
    if( ! clut[i] )
    {
      fprintf(stderr,"malloc failed for clut\n");
      exit(1);
    }
  }
  set_clut(g.n_col,clut[0], clut[1], clut[2]);
 init_graphics(root,window_id, !e_events, request_clear, &g, clut[0], clut[1], clut[2]);

  for(i=0;i<3;i++)
  {
    free(clut[i]);
  }

    
  seed_uni(seed);

  init_artist_variables();
  sa.sa_handler = &finish_prog;
  sigaction(SIGINT, &sa, NULL );
  sigaction(SIGTERM, &sa, NULL );
  sigaction(SIGHUP, &sa, NULL );
  sigaction(SIGQUIT, &sa, NULL );


  /* This is a stand in for the event loop in a Widget set implementation 
   * where we would call plot_column at regular intervals using
   * XTtimeout. However xmountains is an Xlib program
   * so I do the following.
   */
  while( TRUE )
  {
    plot_column(&g);
    zap_events();
#ifndef NO_SLEEP
    if( g.scroll ){
    /* sleep if we are due a scroll next time */
    /* sleeping is very bad because it will prevent
     * events being processed but I suppose it is better
     * than being a CPU hog, as a compremise always check for
     * events at least once a second, looping for longer sleep times.
     * process the events before a sleep to make sure the screen is up to date.
     * the events must always be processed at least once.
     */
      for(i=0;i<snooze;i++){
	sleep(1);
	zap_events();
      }
    }
#endif
  }
}

    
extern int quit_xmount;

void finish_prog(int err)
{
  /* The next time zap_events is called the program will quit */
  quit_xmount=TRUE;
}

